### PR TITLE
Find common ancestors for multiple generators/functions

### PR DESCRIPTION
**Pull request by hermanventer**
Keep track of the parent generator of nested generators/functions. Use this parent edge to compute the common ancestor of a list of generators/functions. Emit declarations that are used in more than one nested scope in their closest common ancestor.

Issue: #482

### CODE REVIEW

File: src/serializer/ResidualHeapSerializer.js [hunk 1]
```diff
@@ -479,29 +481,18 @@ export class ResidualHeapSerializer {
...
-        invariant(body === this.emitter.getBody());
-        return { body };
-      }
+    // This value is referenced from more than one generator or function.
+    // We can emit the initialization of this value into the body associated with their common ancestor.
+    let commonAncestor = scopeList.reduce((x, y) => x && commonAncestorOf(x, y), generators[0]);
+    invariant(commonAncestor instanceof Generator);
+    let body = commonAncestor === this.generator ? this.mainBody : this.activeGeneratorBodies.get(commonAncestor);
+    if (body === undefined) {
+      // TODO: The generator is no longer active! We missed the right time to emit this value.
```
Comment (NTillmann) [inline src/serializer/ResidualHeapSerializer.js]: This remaining TODO is not completely unrelated to the comment about #482 below.

File: src/serializer/ResidualHeapSerializer.js [hunk 2]
```diff
@@ -479,29 +481,18 @@ export class ResidualHeapSerializer {
...
-        // TODO: The generator is no longer active! We missed the right time to emit this value.
-        this.logger.logError(val, "Value is referenced in an unsupported scope.");
-        return { body: this.mainBody };
-      } else {
-        invariant(body === this.emitter.getBody());
-        return { body };
-      }
+    // This value is referenced from more than one generator or function.
+    // We can emit the initialization of this value into the body associated with their common ancestor.
+    let commonAncestor = scopeList.reduce((x, y) => x && commonAncestorOf(x, y), generators[0]);
```
Comment (NTillmann) [inline src/serializer/ResidualHeapSerializer.js]: What does the `x &&` do?

File: src/serializer/ResidualHeapSerializer.js [hunk 3]
```diff
@@ -450,7 +450,9 @@ export class ResidualHeapSerializer {
     // First, let's figure out from which function and generator scopes this value is referenced.
     let functionValues = [];
     let generators = [];
+    let scopeList = [];
```
Comment (NTillmann) [inline src/serializer/ResidualHeapSerializer.js]: This scopeList seems to be unnecessary. Down where you need it, you should be able to just say `Array.from(scopes)`.

**Comment by cblappert** [comment]
What about the case where the common ancestor is a generator that gets thrown away by applyEffects? I think this exhibits it (the common ancestor to declare y is the factory function of module 1), that function body should be executed in an evaluateForEffects context that is then applied.

```
// add require boilerplate here
let a = global.__abstract ? __abstract("boolean", "(false)") : false;
let x = global.__abstract ? __abstract("number", "(42)") : 42;
define(function(global, require, module, exports) {
  module.exports = require(1);
}, 0, null);
define(function(global, require, module, exports) {
   let y = x * 2;
   if (a) {
     global.z = y;
   } else {
     global.z = y;
   }
  module.exports = global.z;
}, 1, null);
require(0);
inspect = function() { return global.z.toString(); }
```

File: src/serializer/ResidualHeapSerializer.js (hunk 3 above)
Comment (hermanventer) [inline src/serializer/ResidualHeapSerializer.js]: will change it.

File: src/serializer/ResidualHeapSerializer.js (hunk 2 above)
Comment (hermanventer) [inline src/serializer/ResidualHeapSerializer.js]: It convinces Flow that x can never be undefined.

File: src/serializer/ResidualHeapSerializer.js (hunk 1 above)
Comment (hermanventer) [inline src/serializer/ResidualHeapSerializer.js]: I'm happy to replace this with an invariant if you believe it should never happen.

**Comment by hermanventer** [comment]
@cblappert: I've added your test case. It turns out that is passes (both in the current form of this request and when stacked on top of the CSE request).

If I understand things correctly, the reason is that the association between value and scope is formed in ResidualHeapVisitor, using only generators that will get emitted. That means the the generators that form part of Effects are unreachable from any serializer code, even after the containing Effects have been applied.

**Comment by hermanventer** [comment]
@NTillmann: I've added your additional test case and it works too. Perhaps it has to be a bit more subtle, because I'm unsurprised that it works.

**Pull request closed without merging**